xmtp_mls/src/groups/mls_sync.rs

```diff
@@ -714,16 +738,45 @@ where
             );
 
             // If no error committing the change, write a transcript message
-            let msg =
-                self.save_transcript_message(validated_commit, envelope_timestamp_ns, *cursor)?;
-            return Ok((IntentState::Committed, msg.map(|m| m.id)));
-        } else if let Some(id) = calculate_message_id_for_intent(intent)? {
-            tracing::debug!("setting message @cursor=[{}] to published", envelope.id);
-            conn.set_delivery_status_to_published(&id, envelope_timestamp_ns, envelope.id as i64)?;
-            return Ok((IntentState::Processed, Some(id)));
+            let msg = self
+                .save_transcript_message(validated_commit, envelope_timestamp_ns, *cursor)
+                .map_err(|err| IntentResolutionError {
+                    processing_error: err,
+                    // If it is a non-retriable error, the commit will be applied, but the transcript message
+                    // will be missing. We mark the intent state as errored and continue.
+                    next_intent_state: IntentState::Error,
+                })?;
+            return Ok(msg.map(|m| m.id));
         }
-        // TODO: Should clarify when this would happen
-        Ok((IntentState::Committed, None))
+
+        let result: Result<Option<Vec<u8>>, IntentError> = calculate_message_id_for_intent(intent);
+        let id = result
+            .map_err(GroupMessageProcessingError::Intent)
+            .transpose()
+            .unwrap_or(Err(GroupMessageProcessingError::MessageIdentifierNotFound))
+            .map_err(|err| {
+                if !err.is_retryable() {
+                    tracing::error!(
+                        "Message identifier not found for intent {} with kind {}, {err:?}",
+                        intent.id,
+                        intent.kind
+                    );
+                }
+                IntentResolutionError {
+                    processing_error: err,
+                    // If the error is non-retriable, it means that the optimistic message (which is already in
+                    // the db) will never have its delivery status updated to published. We mark the intent state
+                    // as errored and continue.
+                    next_intent_state: IntentState::Error,
+                }
+            })?;
+        tracing::debug!("setting message @cursor=[{}] to published", envelope.id);
+        conn.set_delivery_status_to_published(&id, envelope_timestamp_ns, envelope.id as i64)
+            .map_err(|err| IntentResolutionError {
+                processing_error: GroupMessageProcessingError::Db(err),
+                next_intent_state: IntentState::Error,
+            })?;
+        Ok(Some(id))
```

Comment:
Change:
- Previously if we were able to set the delivery status for the message, we marked the intent as `Processed`. Now, we simply return `Ok()`, and the outer function marks the intent as `Committed`.

This undoes the change from https://github.com/xmtp/libxmtp/pull/1899. I compensated by adding the logic on L663, which I believe addresses the original issue.